tests/visualization_tests/test_pareto_front.py

```diff
@@ -1,451 +1,249 @@
 from optuna.trial import TrialState
 from optuna.visualization import plot_pareto_front
+import optuna.visualization._pareto_front
+from optuna.visualization._pareto_front import _get_pareto_front_info
 from optuna.visualization._pareto_front import _make_hovertext
+from optuna.visualization._pareto_front import _ParetoFrontInfo
 from optuna.visualization._plotly_imports import go
 from optuna.visualization._utils import COLOR_SCALE
+from optuna.visualization.matplotlib._matplotlib_imports import plt
+import optuna.visualization.matplotlib._pareto_front
 
 
-def _check_data(figure: "go.Figure", axis: str, expected: Sequence[int]) -> None:
-    """Compare `figure` against `expected`.
+def test_get_pareto_front_info_infer_n_targets() -> None:
+    study = optuna.create_study(directions=["minimize", "minimize"])
+    assert _get_pareto_front_info(study).n_targets == 2
 
-    Concatenate `data` in `figure` in reverse order, pick the desired `axis`, and compare with
-    the `expected` result.
+    study = optuna.create_study(directions=["minimize"] * 5)
+    assert (
+        _get_pareto_front_info(
+            study, target_names=["target1", "target2"], targets=lambda _: [0.0, 1.0]
+        ).n_targets
+        == 2
+    )
 
-    Args:
-        figure: A figure.
-        axis: The axis to be checked.
-        expected: The expected result.
-    """
+    study = optuna.create_study(directions=["minimize"] * 5)
+    study.optimize(lambda _: [0] * 5, n_trials=1)
+    assert _get_pareto_front_info(study, targets=lambda _: [0.0, 1.0]).n_targets == 2
 
-    n_data = len(figure.data)
-    actual = tuple(
-        itertools.chain(*list(map(lambda i: figure.data[i][axis], reversed(range(n_data)))))
-    )
-    assert actual == expected
+    study = optuna.create_study(directions=["minimize"] * 2)
+    with pytest.raises(ValueError):
+        _get_pareto_front_info(study, targets=lambda _: [0.0, 1.0])
 
 
-@pytest.mark.parametrize("include_dominated_trials", [False, True])
-@pytest.mark.parametrize("use_constraints_func", [False, True])
-@pytest.mark.parametrize("axis_order", [None, [0, 1], [1, 0]])
-@pytest.mark.parametrize("targets", [None, lambda t: (t.values[0], t.values[1])])
-def test_plot_pareto_front_2d(
-    include_dominated_trials: bool,
-    use_constraints_func: bool,
-    axis_order: Optional[List[int]],
-    targets: Optional[Callable[[FrozenTrial], Sequence[float]]],
-) -> None:
-    if axis_order is not None and targets is not None:
-        pytest.skip("skip using both axis_order and targets")
+def create_study_2d() -> Study:
 
-    # Test with no trial.
     study = optuna.create_study(directions=["minimize", "minimize"])
-    figure = plot_pareto_front(
-        study=study,
-        include_dominated_trials=include_dominated_trials,
-        axis_order=axis_order,
-    )
-    assert len(figure.data) == 2
-    assert (figure.data[1]["x"] + figure.data[0]["x"]) == ()
-    assert (figure.data[1]["y"] + figure.data[0]["y"]) == ()
```

Comment:
Tests for empty study is needed only because we need to check whether the inference of `n_targets` is working well. (There seems to be no other branches depending on whether the study is empty or not.) We gather that test in `test_get_pareto_front_info_infer_n_targets`.